crud/src/main/java/com/redhat/lightblue/rest/crud/cmd/FindCommand.java

```diff
@@ -155,14 +169,26 @@ public CallStatus run() {
                 streamResponse=getMediator().findAndStream(ireq);
                 return new CallStatus(new Response());
             } else {
-                return new CallStatus(getMediator().find(ireq));
+                r = getMediator().find(ireq);
+                return new CallStatus(r);
             }
         } catch (Error e) {
+            metricCtx.markRequestException(e);
             LOGGER.error("find:generic_error failure: {}", e);
             return new CallStatus(e);
         } catch (Exception e) {
+            Error error = Error.get(RestCrudConstants.ERR_REST_FIND, e.toString());	
+            metricCtx.markRequestException(error);
             LOGGER.error("find:generic_exception failure: {}", e);
-            return new CallStatus(Error.get(RestCrudConstants.ERR_REST_FIND, e.toString()));
+            return new CallStatus(error);
+        } finally {
+            if (!stream) {
```

Comment:
There are failure cases where we are streaming, but the context is not ended. We should unit test these cases to make sure they are caught and handled. For example, the catch block at line 150: a failure there with a streaming request will never close the context.

I think if you make line 185 `if (streamResponse == null) {` instead of `if (!stream) {` it will fix these cases, but I'd highly recommend having unit tests to make sure these branches are all handled as it's too easy to miss one.

Reply:
Using streamresponse null check does handle both streaming and non streaming cases more efficiently. Changed the test condition accordingly and added test cases in FindCommandTest to test both streaming and non streaming cases with metrics handling to cover all cases.